core/server/workspace.go

```diff
@@ -145,12 +147,18 @@ func (ws *Workspace) RefreshAllComponents(ctx context.Context, input *api.Refres
 	if err != nil {
 		return nil, err
 	}
-	// TODO: Parallelism.
-	for _, component := range components.Components {
-		if err := ws.refreshComponent(ctx, ws.Store, component); err != nil {
-			// TODO: Error recovery.
-			return nil, fmt.Errorf("refreshing %q: %w", component.Name, err)
+	if err := ws.TaskTracker.RunFunc(ctx, "refresh", func(refreshTask *task.Task) error {
```

Comment:
I will probably replace this usage of `RunFunc` with `Start/Wait` pairs, so that the return value handling is less ugly & makes less use of higher-order programming. This was just my first attempt at getting this to work.